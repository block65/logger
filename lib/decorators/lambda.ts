import { Decorator } from '../logger.js';

export const lambdaDecorator: Decorator = (log) => {
  return {
    ...log,
  };
};
