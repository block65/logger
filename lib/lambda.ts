import type { Context } from 'aws-lambda';
import type { JsonObject } from 'type-fest';
import type { Logger } from './logger.js';
import { isEmptyObject } from './utils.js';

type LambdaContext = Partial<Context> & Required<Pick<Context, 'awsRequestId'>>;

export function withLambdaLoggerContextWrapper<T>(
  logger: Logger,
  lambdaContext: LambdaContext,
  fn: (...args: any) => Promise<T>,
): Promise<T> {
  const context: JsonObject = {
    // functionName: context.functionName,
    ...(lambdaContext.functionVersion !== '$LATEST' && {
      functionVersion: lambdaContext.functionVersion,
    }),
  };

  return logger.als.run(
    {
      id: lambdaContext.awsRequestId,
      ...(!isEmptyObject(context) && { context }),
    },
    () => {
      return fn();
    },
  );
}
