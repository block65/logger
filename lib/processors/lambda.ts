import { Processor } from '../logger.js';

export const lambdaProcessor: Processor = (log) => {
  return {
    ...log,
  };
};
