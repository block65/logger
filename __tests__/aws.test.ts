import {
  afterAll,
  afterEach,
  beforeEach,
  describe,
  expect,
  jest,
  test,
} from '@jest/globals';
import { lambdaLoggerContextWrapper } from '../lib/lambda.js';
import { createLoggerWithWaitableMock } from './helpers.js';

describe('AWS', () => {
  beforeEach(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('2009-02-13T23:31:30.000Z'));
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  test('Lambda Logger + contextId', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock();

    const contextId = 'fake-aws-request-id';

    const closure = lambdaLoggerContextWrapper(logger, contextId, {
      stuff: true,
    });

    expect.assertions(1);

    await new Promise<void>((resolve) => {
      closure(async () => {
        logger.warn('hello');
        await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
        resolve();
      });
    });
  });

  test('Lambda Error Object', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock({
      platform: 'aws-lambda',
    });

    logger.error(new Error('Ded 4'));
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });

  test('Error Object', async () => {
    const [logger, callback] = await createLoggerWithWaitableMock({
      platform: 'aws',
    });

    logger.error(new Error('Ded 3'));
    await expect(callback.waitUntilCalled()).resolves.toMatchSnapshot();
  });
});
