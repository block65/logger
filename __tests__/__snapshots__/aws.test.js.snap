// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AWS ECS Error Object 1`] = `
Array [
  Array [
    "2009-02-13T23:31:30.000Z		ERROR	Ded 3	{\\"level\\":50,\\"err\\":{\\"name\\":\\"Error\\",\\"message\\":\\"Ded 3\\",\\"stack\\":\\"Error: Ded 3\\\\n    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/aws.test.js:51:26)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\\"}}
",
  ],
]
`;

exports[`AWS Lambda Error Object 1`] = `
Array [
  "2009-02-13T23:31:30.000Z		ERROR	Ded 4	{\\"level\\":50,\\"err\\":{\\"name\\":\\"Error\\",\\"message\\":\\"Ded 4\\",\\"stack\\":\\"Error: Ded 4\\\\n    at Object.<anonymous> (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/__tests__/aws.test.js:41:26)\\\\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\\"}}
",
]
`;

exports[`AWS Lambda Logger + contextId 1`] = `
Object {
  "somethingUseful": true,
}
`;

exports[`AWS Lambda Logger + contextId 2`] = `
Array [
  Array [
    "2009-02-13T23:31:30.000Z		WARN	hello	{\\"level\\":40}
",
  ],
  Array [
    "2009-02-13T23:31:30.000Z		ERROR	fake	{\\"level\\":50,\\"err\\":{\\"name\\":\\"Error\\",\\"message\\":\\"fake\\",\\"stack\\":\\"Error: fake\\\\n    at fn (~/block65_logger/bazel-out/k8-fastbuild/~/block65_logger/lib/lambda.js:12:16\\\\n    at AsyncLocalStorage.run (node:async_hooks:320:14)\\"}}
",
  ],
]
`;
